barretenberg/cpp/src/barretenberg/sumcheck/sumcheck.hpp

```diff
@@ -178,57 +199,97 @@ template <typename Flavor> class SumcheckProver {
+                                                        zk_sumcheck_data);
+            // Place evaluations of Sumcheck Round Univariate in the transcript
             transcript->send_to_verifier("Sumcheck:univariate_" + std::to_string(round_idx), round_univariate);
             FF round_challenge = transcript->template get_challenge<FF>("Sumcheck:u_" + std::to_string(round_idx));
             multivariate_challenge.emplace_back(round_challenge);
+            // Prepare sumcheck book-keeping table for the next round
             partially_evaluate(partially_evaluated_polynomials, round.round_size, round_challenge);
+            // Prepare evaluation masking and libra structures for the next round (for ZK Flavors)
+            if constexpr (Flavor::HasZK) {
+                update_zk_sumcheck_data(zk_sumcheck_data, round_challenge, round_idx);
+            };
+
             pow_univariate.partially_evaluate(round_challenge);
             round.round_size = round.round_size >> 1;
         }
+        // Check that the challenges \f$ u_0,\ldots, u_{d-1} \f$ do not satisfy the equation \f$ u_0(1-u_0) + \ldots +
+        // u_{d-1} (1 - u_{d-1}) = 0 \f$. This equation is satisfied with probability ~ 1/|FF|, in such cases the prover
+        // has to abort and start ZK Sumcheck anew.
+        if constexpr (Flavor::HasZK) {
+            check_that_evals_do_not_leak_witness_data(multivariate_challenge);
+        };
+        // Zero univariates are used to pad the proof to the fixed size CONST_PROOF_SIZE_LOG_N.
         auto zero_univariate = bb::Univariate<FF, Flavor::BATCHED_RELATION_PARTIAL_LENGTH>::zero();
         for (size_t idx = multivariate_d; idx < CONST_PROOF_SIZE_LOG_N; idx++) {
             transcript->send_to_verifier("Sumcheck:univariate_" + std::to_string(idx), zero_univariate);
             FF round_challenge = transcript->template get_challenge<FF>("Sumcheck:u_" + std::to_string(idx));
             multivariate_challenge.emplace_back(round_challenge);
         }
-
-        // Final round: Extract multivariate evaluations from #partially_evaluated_polynomials and add to transcript
+        // The evaluations of Libra uninvariates at \f$ g_0(u_0), \ldots, g_{d-1} (u_{d-1}) \f$ are added to the
+        // transcript.
+        if constexpr (Flavor::HasZK) {
+            transcript->send_to_verifier("Libra:evaluations", zk_sumcheck_data.libra_evaluations);
+        };
+
+        // Claimed evaluations of Prover polynomials are extracted and added to the transcript. When Flavor has ZK, the
+        // evaluations of all witnesses are masked.
         ClaimedEvaluations multivariate_evaluations;
-        for (auto [eval, poly] :
-             zip_view(multivariate_evaluations.get_all(), partially_evaluated_polynomials.get_all())) {
-            eval = poly[0];
-        }
-        transcript->send_to_verifier("Sumcheck:evaluations", multivariate_evaluations.get_all());
+        if constexpr (Flavor::HasZK) {
```

Comment:
I removed the second method, so now extract claimed evaluations accepts zk_sumcheck_data optionally. As with extend_edges, I decided to keep the if here, so that it's clear that in non-zk cases nothing is passed to it.